### PR TITLE
Very promising apis for swift :)

### PR DESCRIPTION
**Pull request by trooper2013**
Still WIP.  First subset of api(s) wrapped in promises. Covers basic SalesforceSDKManager, SFUserAccountManager & RestAPI calls.  NOTE: The tests are functional  but needs better setup code. I will revamp it in subsequent PR's. Some examples of API(s) 
```
 SFUserAccountManager.sharedInstance().Promises
            .login()
            .then { userAccount  in
               ....
            }
```

```
let restApi  = SFRestAPI.sharedInstance()
restApi.Factory.query(soql: "SELECT Id,FirstName,LastName FROM User")
 .then { request in
     restApi.send(request: request)
}
 .done { data in
     var  restResonse = data.asJsonDictionary()  
     ...      
}
 .catch {  error in
             
 }
```

### CODE REVIEW

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Classes/SalesforceSwiftLogger.swift [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /*
- SalesforceSDKCoreExtensions
- Created by Raj Rao on 11/27/17.
+ Logger
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Classes/SalesforceSwiftLogger.swift]: `SalesforceSwiftLogger`.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+         SFRestRequestFactory.Factory.resources()
+         .then { (request) in
+             restApi.send(request)
+         }
+         ```
+         - Returns: SFRestRequest wrapped in a promise.
+         */
+        public func resources() -> Promise<SFRestRequest> {
+            return  Promise(.pending) {  resolver in
+                resolver.fulfill(self.api!.requestForVersions())
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift]: Should be `requestForResources()` -> copy-paste error?

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+         SFRestRequestFactory.Factory.describe()
+         .then { (request) in
+             restApi.send(request)
+         }
+         ```
+         - Returns: SFRestRequest wrapped in a promise.
+         */
+        public func describe(objectType:String) -> Promise<SFRestRequest> {
+            return  Promise(.pending) {  resolver in
+                resolver.fulfill(self.api!.requestForVersions())
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift]: Should be `requestForDescribe()` -> copy-paste error?

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        public func delete(objectType: String, objectId: String) -> Promise<SFRestRequest> {
+            return  Promise(.pending) {  resolver in
+                resolver.fulfill(self.api!.requestForDelete(withObjectType: objectType, objectId: objectId))
+            }
+        }
+        
+        /**
+         A factory method for update object request.
+         ```
+         SFRestRequestFactory.Factory.describeGlobal()
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift]: `query()`.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        public func query(soql: String) -> Promise<SFRestRequest> {
+            return  Promise(.pending) {  resolver in
+                resolver.fulfill(self.api!.request(forQuery: soql))
+            }
+        }
+        
+        /**
+         A factory method for update object request.
+         ```
+         SFRestRequestFactory.Factory.describeGlobal()
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift]: `queryAll()`.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFUserAccountManagerExtensions.swift [hunk 6]
```diff
@@ -0,0 +1,190 @@
+/*
+ SFRestAPIExtensions
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFUserAccountManagerExtensions.swift]: `SFUserAccountManagerExtensions`.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SalesforceSDKManagerExtensions.swift [hunk 7]
```diff
@@ -0,0 +1,166 @@
+/*
+ SalesforceSDKCoreExtensions
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SalesforceSDKManagerExtensions.swift]: `SalesforceSDKManagerExtensions`.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SwiftExtensions.swift [hunk 8]
```diff
@@ -0,0 +1,79 @@
+/*
+ SalesforceSDKCoreExtensions
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SwiftExtensions.swift]: `SwiftExtensions`.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SalesforceTestExtensions.swift [hunk 9]
```diff
@@ -0,0 +1,198 @@
+/*
+ SalesforceSwiftSDKTests
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SalesforceTestExtensions.swift]: `SalesforceTestExtensions`.

**Comment by bhariharan** [comment]
@trooper2013 You need to merge `dev` and resolve conflicts too.

**Comment by brandonpage** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Big PR, try to keep changes smaller if you can</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Needs testing on a Phone if change is non-trivial</td>
    </tr>
  </tbody>
</table>

## Current coverage for SalesforceAnalytics.framework is `66.64%`
✅ *No files affecting coverage found*

---
## Current coverage for SalesforceHybridSDK.framework is `0.00%`
✅ *No files affecting coverage found*

---
## Current coverage for SalesforceSDKCore.framework is `50.64%`
✅ *No files affecting coverage found*

---
## Current coverage for SmartStore.framework is `62.81%`
✅ *No files affecting coverage found*

---
## Current coverage for SmartSync.framework is `78.88%`
✅ *No files affecting coverage found*

---

> Powered by [xcov](https://github.com/nakiostudio/xcov)

<p align="right" data-meta="generated_by_CI-result">
  Generated by :no_entry_sign: <a href="http://danger.systems/">Danger</a>
</p>

**Comment by trooper2013** [comment]
@wmathurin The work is not complete. The rest of them will be in the next PR's.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+        public func metadata(objectType: String) -> Promise<SFRestRequest> {
+            return  Promise(.pending) {  resolver in
+                resolver.fulfill(self.api!.requestForMetadata(withObjectType: objectType))
+            }
+        }
+        
+        /**
+         A factory method for retrieve object request.
+         ```
+         SFRestRequestFactory.Factory.describeGlobal()
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift]: Should be `retrieve`.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+         */
+        public func update(objectType: String,objectId: String,fieldList: [String: Any]?,ifUnmodifiedSince: Date?) -> Promise<SFRestRequest> {
+            return  Promise(.pending) {  resolver in
+                resolver.fulfill(self.api!.requestForUpdate(withObjectType: objectType, objectId: objectId, fields: fieldList, ifUnmodifiedSince: ifUnmodifiedSince))
+            }
+        }
+      
+        
+        /**
+         A factory method for update object request.
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift]: Should be `delete`.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+         - Returns:  SFRestRequest wrapped in a promise.
+         */
+        public func delete(objectType: String, objectId: String) -> Promise<SFRestRequest> {
+            return  Promise(.pending) {  resolver in
+                resolver.fulfill(self.api!.requestForDelete(withObjectType: objectType, objectId: objectId))
+            }
+        }
+        
+        /**
+         A factory method for update object request.
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift]: Should be `query`.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+         - Returns:  SFRestRequest wrapped in a promise.
+         */
+        public func query(soql: String) -> Promise<SFRestRequest> {
+            return  Promise(.pending) {  resolver in
+                resolver.fulfill(self.api!.request(forQuery: soql))
+            }
+        }
+        
+        /**
+         A factory method for update object request.
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDK/Extensions/SFRestAPIExtensions.swift]: Should be `queryAll`.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SFRestAPITests.swift [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+                SFUserAccountManager.sharedInstance().currentUser = userAccount
+                setupComplete = true
+            }.catch { error  in
+                setupComplete = true
+        }
+    }
+    
+    override func setUp() {
+        super.setUp()
+        //let exp1 = expectation(description: "init")
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SFRestAPITests.swift]: Remove commented out code.

File: libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SFRestAPITests.swift [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+        SalesforceSwiftSDKTests.waitForCompletion(maxWaitTime: 5) { () -> Bool in
+            if (SalesforceSwiftSDKTests.setupComplete) {
+                return true
+            }
+            return false
+        }
+        //  waitForExpectations(timeout: 10)
```
Comment (bhariharan) [inline libs/SalesforceSwiftSDK/SalesforceSwiftSDKTests/SFRestAPITests.swift]: Remove commented out code.

**Pull request merged**